File: bindings/python/tests/test_kem.py
# libpqc-dyber - Post-Quantum Cryptography Library
# Copyright (c) 2024-2026 Dyber, Inc.
# SPDX-License-Identifier: Apache-2.0 OR MIT

"""Pytest tests for the KEM binding."""

import pytest

from pqc_dyber import KEM, AlgorithmNotSupportedError, PQCError, kem_algorithms, kem_is_enabled


class TestKEMProperties:
    """Test KEM context properties."""

    def test_algorithm_name(self):
        kem = KEM("ML-KEM-768")
        assert kem.algorithm == "ML-KEM-768"

    def test_sizes_are_positive(self):
        kem = KEM("ML-KEM-768")
        assert kem.public_key_size > 0
        assert kem.secret_key_size > 0
        assert kem.ciphertext_size > 0
        assert kem.shared_secret_size > 0

    def test_repr(self):
        kem = KEM("ML-KEM-768")
        assert repr(kem) == "KEM('ML-KEM-768')"

    def test_invalid_algorithm_raises(self):
        with pytest.raises(AlgorithmNotSupportedError):
            KEM("INVALID-ALGORITHM-NAME")


class TestKEMOperations:
    """Test KEM keygen / encaps / decaps round-trip."""

    def test_roundtrip_mlkem768(self):
        kem = KEM("ML-KEM-768")
        pk, sk = kem.keygen()

        assert isinstance(pk, bytes)
        assert isinstance(sk, bytes)
        assert len(pk) == kem.public_key_size
        assert len(sk) == kem.secret_key_size

        ct, ss_enc = kem.encaps(pk)
        assert isinstance(ct, bytes)
        assert isinstance(ss_enc, bytes)
        assert len(ct) == kem.ciphertext_size
        assert len(ss_enc) == kem.shared_secret_size

        ss_dec = kem.decaps(ct, sk)
        assert isinstance(ss_dec, bytes)
        assert len(ss_dec) == kem.shared_secret_size
        assert ss_enc == ss_dec

    def test_roundtrip_mlkem512(self):
        kem = KEM("ML-KEM-512")
        pk, sk = kem.keygen()
        ct, ss_enc = kem.encaps(pk)
        ss_dec = kem.decaps(ct, sk)
        assert ss_enc == ss_dec

    def test_roundtrip_mlkem1024(self):
        kem = KEM("ML-KEM-1024")
        pk, sk = kem.keygen()
        ct, ss_enc = kem.encaps(pk)
        ss_dec = kem.decaps(ct, sk)
        assert ss_enc == ss_dec

    def test_different_keypairs_produce_different_keys(self):
        kem = KEM("ML-KEM-768")
        pk1, sk1 = kem.keygen()
        pk2, sk2 = kem.keygen()
        # Keys should differ (with overwhelming probability).
        assert pk1 != pk2
        assert sk1 != sk2


class TestKEMAlgorithmEnumeration:
    """Test algorithm listing and query functions."""

    def test_kem_algorithms_not_empty(self):
        algs = kem_algorithms()
        assert len(algs) > 0

    def test_mlkem768_is_listed(self):
        algs = kem_algorithms()
        assert "ML-KEM-768" in algs

    def test_kem_is_enabled(self):
        assert kem_is_enabled("ML-KEM-768") is True
        assert kem_is_enabled("NONEXISTENT") is False

File: bindings/rust/tests/test_kem.rs
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT
//
// Integration tests for the KEM wrapper.

use pqc_dyber::{init, cleanup, Kem};

fn setup() {
    let _ = init();
}

#[test]
fn test_kem_mlkem768_roundtrip() {
    setup();

    let kem = Kem::new("ML-KEM-768").expect("failed to create ML-KEM-768 context");
    assert_eq!(kem.algorithm(), "ML-KEM-768");

    let (pk, sk) = kem.keygen().expect("keygen failed");
    assert_eq!(pk.len(), kem.public_key_size());
    assert_eq!(sk.len(), kem.secret_key_size());

    let (ct, ss_enc) = kem.encaps(&pk).expect("encaps failed");
    assert_eq!(ct.len(), kem.ciphertext_size());
    assert_eq!(ss_enc.len(), kem.shared_secret_size());

    let ss_dec = kem.decaps(&ct, &sk).expect("decaps failed");
    assert_eq!(ss_enc, ss_dec);

    cleanup();
}

#[test]
fn test_kem_mlkem512_roundtrip() {
    setup();

    let kem = Kem::new("ML-KEM-512").expect("failed to create ML-KEM-512 context");

    let (pk, sk) = kem.keygen().expect("keygen failed");
    let (ct, ss_enc) = kem.encaps(&pk).expect("encaps failed");
    let ss_dec = kem.decaps(&ct, &sk).expect("decaps failed");

    assert_eq!(ss_enc, ss_dec);

    cleanup();
}

#[test]
fn test_kem_mlkem1024_roundtrip() {
    setup();

    let kem = Kem::new("ML-KEM-1024").expect("failed to create ML-KEM-1024 context");

    let (pk, sk) = kem.keygen().expect("keygen failed");
    let (ct, ss_enc) = kem.encaps(&pk).expect("encaps failed");
    let ss_dec = kem.decaps(&ct, &sk).expect("decaps failed");

    assert_eq!(ss_enc, ss_dec);

    cleanup();
}

#[test]
fn test_kem_unsupported_algorithm() {
    setup();
    let result = Kem::new("NONEXISTENT-KEM");
    assert!(result.is_err());
    cleanup();
}

#[test]
fn test_kem_wrong_key_size() {
    setup();

    let kem = Kem::new("ML-KEM-768").expect("failed to create KEM context");
    let bad_pk = vec![0u8; 10]; // wrong size

    let result = kem.encaps(&bad_pk);
    assert!(result.is_err());

    cleanup();
}

#[test]
fn test_kem_security_level() {
    setup();

    let kem = Kem::new("ML-KEM-768").expect("ML-KEM-768");
    assert!(kem.security_level() >= 1 && kem.security_level() <= 5);

    cleanup();
}

#[test]
fn test_kem_algorithm_enumeration() {
    setup();

    let count = pqc_dyber::kem_algorithm_count();
    assert!(count > 0, "should have at least one KEM algorithm");

    for i in 0..count {
        let name = pqc_dyber::kem_algorithm_name(i);
        assert!(name.is_some(), "algorithm index {} should have a name", i);
    }

    cleanup();
}
